CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form ("view mode" for questionnaires)

veronikaslc opened the pull request:
Contains:
[CARDS-1268](https://phenotips.atlassian.net/browse/CARDS-1268) Refactor FormPagination
[CARDS-1269](https://phenotips.atlassian.net/browse/CARDS-1269) QuestionnairePreview component
[CARDS-1270](https://phenotips.atlassian.net/browse/CARDS-1270) Menu for the Questionnare page (edit mode)

To test:
* The questionnaire preview in the administration - ensure that all types of fields work as they would in a form
* The pagination on both forms and questionnaire preview

Commit:
CARDS-1268: Refactor FormPagination

Commit:
CARDS-1269: QuestionnairePreview component

Commit:
CARDS-1270: Menu for the Questionnare page (edit mode)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -66,9 +146,10 @@ function FormPagination (props) {
       onClick={handleNext}
     >
-      {((lastPage() === 0 || activePage === lastPage()) && saveInProgress) ? 'Saving' :
+      {!enableSave ? "Next" :
```
When `enableSave` is `false`, the Next button should be disabled on the last page.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -17,28 +17,95 @@
  */
 function FormPagination (props) {
-  let { classes, lastPage, activePage, saveInProgress, lastSaveStatus, handlePageChange } = props;
+  let { classes, saveInProgress, lastSaveStatus, setPagesCallback, paginationEnabled, enableSave, questionnaireData } = props;
```
Do we need a `paginationEnabled` prop? Wouldn't the pagination component just not be rendered by the Form/Questionnaire preview if pagination was not enabled?

veronikaslc replied:
Yes, we do. In that case when pagination is not enabled we still need to run component to generate pages.

marta- reviewed (commented):
Looks great so far. A couple of issues:
* **Usability**: When creating a new Questionnaire, it should be opened in edit mode (currently after creation the user is directed to the preview mode)
* **Usability**: On paginated questionnaires.`Next` should appear disabled on the questionnaire preview's last page
* **Usability**: When loading the questionnaire preview, an "Error" message is briefly displayed. Should be a circular progress instead.
* **Bug**: Pedigree and File questions (which currently require form saving to work) have buggy behaviors in the questionnaire preview: can't set a pedigree answer, uploading a file leaves the question in a 'loading state', and booth display console errors saying `saveForm` is not a function. There should be a mock `saveForm` in the questionnaire preview allowing these questions to function properly.

sashaandjic commented:
Confirming Marta's findings

Commit:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form (view mode for questionnaires)
Addressed codereview comments

sashaandjic commented:
Pagination still does not work. Here is the look of the Patient form that has `Paginate` = `Yes`.
SAVE button is disabled, however there is no NEXT button nor fields are grouped into pages.
![image](https://user-images.githubusercontent.com/23243244/128815227-98831e1a-e7e8-4b54-b466-92e9d0a9bf21.png)

All other problems, previously reported, are fixed.

marta- commented:
I confirm that pagination no longer works on forms. It works as expected on the questionnaire preview.

marta- commented:
> * File type questions: works better than the previous iteration, the file(s) appear to be uploaded like they would on a form, but [the progress bar never disappears](https://user-images.githubusercontent.com/651980/128879582-5e3399ab-05f4-49da-bb60-539479daf6bc.png). Is it possible for the mock `saveForm` to return something that will integrate with the File question functionality and make the progress bar disappear?

Also because of the way file questions are implemented, when attempting to test a File question in the questionnaire preview, the file is actually uploaded to that question node. This should be prevented. I see two options:
1. somehow disable "answering" File questions in preview mode
2.  update the Form component and the File question (and the Questionnaire preview) so that a method from the form context is used for uploading files, instead of performing the direct upload

Commit:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form (view mode for questionnaires)
Addressed codereview comments: disabled file upload for a view mode; fixed forms pagination

sashaandjic commented:
The only issue that I ran into is with the file types. 
In preview mode, when I drag the file over , the progress bar seems to not go away. 
Everything else works fine.

Commit:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form (view mode for questionnaires)
disabled file upload for a view mode

veronikaslc commented:
@sashaandjic Should be fixed / disabled now.

sashaandjic reviewed (approved):
tested

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -57,18 +124,32 @@ function FormPagination (props) {
       variant="contained"
       color="primary"
-      disabled={saveInProgress}
+      disabled={saveInProgress || activePage === lastValidPage()}
```
Bug: This makes the save button always disabled on paginated forms.
```suggestion
      disabled={saveInProgress || !enableSave && activePage === lastValidPage()}
```

marta- replied:
Fixed in bc306fc .

marta- commented on modules/commons/src/main/frontend/src/components/dragAndDrop.jsx:
```diff
@@ -56,7 +56,7 @@ const useStyles = makeStyles(theme => ({
 
 export default function DragAndDrop(props) {
-  const { accept, multifile, handleDrop, error } = props;
+  const { accept, multifile, handleDrop, error, isDemo } = props;
```
`isDemo` is too specific. It should be renamed to  `disabled`, as there may be other reasons why we would need to disable drag&drop.

marta- replied:
Done in b2cb137 .

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -290,6 +290,7 @@ function FileResourceQuestion(props) {
         handleDrop={addFiles}
         multifile={maxAnswers != 1}
         error={error}
+        isDemo={window.location.pathname.startsWith("/content.html/admin/Questionnaires/")}
```
I disagree with this method for determining whether the upload functionality should be disabled, it's very brittle. It should not be this component's responsibility to determine where the request is coming from and modify its behavior depending on it. Instead, I suggest we enable/disable uploads via the context provider.

marta- replied:
Done in b2cb137 .

marta- committed:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form (view mode for questionnaires)
Fixed bug: save button always disabled on paginated forms

marta- committed:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form (view mode for questionnaires)
Previewing questionnaires with file questions:
* Renamed DragAndDrop parameter from isDemo to disabled
* Prevented progress bar from appearing when dropping files on a disabled DragAndDrop
* Disabled file question via context provider

marta- commented:
> The only issue that I ran into is with the file types.
> In preview mode, when I drag the file over , the progress bar seems to not go away.

Fixed in b2cb137 .

sashaandjic commented:
@marta- no functional changes? I already tested change made in [`b2cb137`](https://github.com/data-team-uhn/cards/commit/b2cb137fe6bbd33dff76a12d9f5e627724404820)

marta- commented:
> @marta- no functional changes? I already tested change made in [`b2cb137`](https://github.com/data-team-uhn/cards/commit/b2cb137fe6bbd33dff76a12d9f5e627724404820)

No functional changes, only the two bugfixes (disabled "save" button on pagination form, and never ending progress bar on file questions preview) + some refactoring and rebasing on top of the latest `dev` branch.

marta- committed:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form ("view mode" for questionnaires)
Removed unnecessary check for disabled status

sashaandjic commented:
2 problems found. Steps:

Create a brand new form with:
- section containing one `info `box, one `file` field (no min, no max) and one `text `field
- another `file `field with min=max=1 outside the section

Create a new questionnaire. 

**Problem 1**: Adding multiple files to the the first `file` type will bring the login dialog ( _please note if you first add only 1 file and then many, everything is good_)

**Problem 2**: Go back to the questionnaire designer and try to delete either `file` or `text `field from the section. The login dialog opens up. Please note that the `info `box and the other `file` type, outside the section, can be deleted with no problem.

![image](https://user-images.githubusercontent.com/23243244/129429869-0bde3f4f-7bc9-4b82-885c-16f4bb583cc3.png)



[json.txt](https://github.com/data-team-uhn/cards/files/6985704/json.txt)

sashaandjic reviewed (approved):
tested

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -17,28 +17,95 @@
+  let addPage = (entryDefinition) => {
+    if (paginationEnabled) {
+      let page;
+      if (QUESTION_TYPES.includes(entryDefinition["jcr:primaryType"]) && previousEntryType && QUESTION_TYPES.includes(previousEntryType)) {
```
Is this line going to have any issues with the newly introduced `cards:Information` primaryType?

sdumitriu replied:
It kind of does. As I understand the code, this is manually grouping consecutive questions that are not part of a section into a separate page, so for example a list of children of type `Q Q Q S Q Q Q` would become 3 pages, one with the first three questions, one with the section in the middle, and one with the last 3 questions. Information nodes would instead become their own pages, which I don't think is what we want.

This is not a targeted usecase; paginated questionnaires should have only sections as direct children. But since we already kind of support it, we should indeed also treat Information nodes as questions, they should not become pages.

marta- replied:
Addressed in edec83a2.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -117,32 +124,70 @@ let Questionnaire = (props) => {
+  let questionnaireMenu = (
+      <div className={classes.actionsMenu}>
+        { isEdit ?
+          <Tooltip title="Preview" onClick={() => setIsEdit(false)}>
```
I don't like that we just change the state and not the actual URL; the URL should reflect the current state.

Possibly related issue: open a questionnaire, edit it, change a question, preview -> still shows the old state, not the updated question; worse: edit again, old state is shown.

marta- replied:
I confirm the old state issue is related. Fixed in 717ccab.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -117,32 +124,70 @@ let Questionnaire = (props) => {
+            </IconButton>
+          </Tooltip>
+        }
+        <DeleteButton
```
Issue: After deleting a questionnaire, the browser stays on the same page; it should go somewhere else (the questionnaires table maybe?)

marta- replied:
Fixed in 717ccab.

sdumitriu reviewed (changes requested):
- opening/closing the questionnaire preview still sends checkout/checkin requests, which is wrong
- closing the pedigree editor should still show the drawn pedigree in the form, even if it is not saved
- paginated form previews: I'm not entirely sure that a disabled "Next" button is good enough on the last page, though I don't have any better alternatives; maybe figure something out if/when we rename the Save button on the last page in the actual form

marta- commented:
> * opening/closing the questionnaire preview still sends checkout/checkin requests, which is wrong
Fixed in 717ccab.

marta- committed:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form ("view mode" for questionnaires)
Addressing review coments:
* no more checkout/checkin in preview mode
* switching between preview and edit now changes the URL, not only the state
  * consequence: preview now always displayes the latest state
* after deleting a quesitonnaire the user is redirected to the Questionnaires table

marta- commented:
> * paginated form previews: I'm not entirely sure that a disabled "Next" button is good enough on the last page, though I don't have any better alternatives; maybe figure something out if/when we rename the Save button on the last page in the actual form

I think it's ok like this for now until we rename the save button.

marta- committed:
CARDS-1303: Add the ability to display informative text in questionnaires
Information cards not included in sections should behave like question cards in the paginated forms

marta- committed:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form ("view mode" for questionnaires)
Quick fix for js console warning displayed when the questionnaire data isn't loaded yet

marta- committed:
CARDS-1263: As an admin, I can access a preview of what a questionnaire looks and functions like without creating a new form ("view mode" for questionnaires)
Addressing review comments: quick fix to allow previewing pedigree questions on the questionnaire page

marta- commented:
> * closing the pedigree editor should still show the drawn pedigree in the form, even if it is not saved

Done in ccf5bd35 .

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/129669292-4aafdea4-6033-4c8e-8be3-58b40bad5812.png)

sashaandjic reviewed (approved):
tested

sashaandjic commented:
Small problem found: On a paginated Patient Info form, closing pedigree editor takes you back to the 1st page instead of 3rd.

QUESTION: We do not update `jcr:lastModified` for Questionnaires? It always equals `jcr:created`?
![image](https://user-images.githubusercontent.com/23243244/129669810-2e5e4d32-f7c8-4218-ad42-ba505645b9eb.png)

marta- commented:
> QUESTION 1: On a paginated Patient Info questionnaire, closing pedigree editor takes you back to the 1st page instead of 2nd. Is it out of scope of this PR as it is happening in the Patient Info form too?

This is https://phenotips.atlassian.net/browse/CARDS-1329, indeed out of scope.
